docs: add electron app page (#259)
casdoor/casdoor-website: docs/how-to-connect/desktop-sdks/electron-app.md

---
title: Electron App
description: An Electron app example for Casdoor
keywords: [electron, sdk]

> ---

Reviewer: ```suggestion
---

```

please leave a blank line in the doc header

An [Electron app example](https://github.com/casdoor/casdoor-electron-example) for Casdoor.

# How to run example

## Init the example

> You need to init requires 6 parameters, which are all string type:

Reviewer: Is this a grammatically wrong sentence?

| redirectPath         | the path of the redirect URL for your Casdoor application, will be `/callback` if not provided | `src/App.js`         |
| clientSecret         | the Client Secret of your Casdoor application                                                   | `src/App.js`         |
| casdoorServiceDomain | your Casdoor server URL                                                                          | `public/electron.js` |

If you don't set these parameters, this project will use the [Casdoor online demo](https://door.casdoor.com/) as the defult Casdoor server and use the [Casnode](https://door.casdoor.com/applications/app-casnode) as the default Casdoor application.

## Available Scripts

In the project directory, you can run:

### `npm run dev` or `yarn dev`

Builds the electron app and run this app.

### `npm run make` or `yarn make`

Package and distribute your application. It will create the `out` folder where your package will be located:
`// Example for macOS out/  
├── out/make/zip/darwin/x64/casdoor-electron-example-darwin-x64-1.0.0.zip  
├── ...  

> └── out/casdoor-electron-example-darwin-x64/casdoor-electron-example.app/Contents/MacOS/casdoor-electron-example`

Reviewer: ![image](https://user-images.githubusercontent.com/53544726/181399811-b5eca34c-36a2-47ef-b37d-1da24b487da8.png)

it doesn't look good.

maybe we should use code block
````md
```
// Example for macOS out/  
├── out/make/zip/darwin/x64/casdoor-electron-example-darwin-x64-1.0.0.zip  
├── ...  
└── out/casdoor-electron-example-darwin-x64/casdoor-electron-example.app/Contents/MacOS/casdoor-electron-example
```
````

## Prview the example

After you run this electron application, a new window will be showed on your desktop.

> ![electron login](/img/howto-desktop-electron-login.png "Image not found: /img/howto-desktop-electron-login.png")

Reviewer: `"Image not found: /img/howto-desktop-electron-login.png"`

what's the use of it ?

If you click `Login with Casdoor` botton, your default browser will be opened automatically and show the login page.
![browser](/img/howto-desktop-electron-browser.png)
After you login successfully, your electron application will be opened and your user name will be showed on your application.
![electron logout](/img/howto-desktop-electron-logout.png)

# How to integrate

> ## Set the custom protocal

Reviewer: ```suggestion
## Set the custom protocol
```

> Firstly, you need to set the custum protocal called `casdoor`.

Reviewer: ```suggestion
Firstly, you need to set the custom protocol called `casdoor`.
```

    const isProtocol = ProtocolRegExp.test(openUrl);
    if (isProtocol) {
      const params = url.parse(openUrl, true).query;
      if (params && params.code) {
        store.set("casdoor_code", params.code);
        mainWindow.webContents.send("receiveCode", params.code);
      }
    }
  });
}
```

You can get the code from the broswer, which is `casdoor_code` or `params.code`.

## Parse the code and get the user info

```javascript
async function getUserInfo(clientId, clientSecret, code) {
  const { data } = await axios({
    method: "post",
    url: authCodeUrl,
    headers: {
      "content-type": "application/json",
    },
    data: JSON.stringify({
      grant_type: "authorization_code",
      client_id: clientId,
      client_secret: clientSecret,
      code: code,
    }),
  });
  const resp = await axios({
    method: "get",
    url: `${getUserInfoUrl}?accessToken=${data.access_token}`,
  });
  return resp.data;
}

ipcMain.handle("getUserInfo", async (event, clientId, clientSecret) => {
  const code = store.get("casdoor_code");
  const userInfo = await getUserInfo(clientId, clientSecret, code);
  store.set("userInfo", userInfo);
  return userInfo;
});
```

> Finally, you can parse the code and get the user info just followed the[ Casdoor docs](http://localhost:3000/docs/how-to-connect/oauth).

Reviewer: ```suggestion
Finally, you can parse the code and get the user info just followed the[ Casdoor docs](/docs/how-to-connect/oauth).
```
